src/sentry/snuba/metrics/metric_extraction.py

```diff
@@ -0,0 +1,258 @@
+    "avg": "d",
+    "quantile(0.50)": "d",
+    "quantile(0.75)": "d",
+    "quantile(0.95)": "d",
+    "quantile(0.99)": "d",
+    "max": "d",
+    # not supported yet
+    "percentile": None,
+    "failure_rate": None,
+    "apdex": None,
+}
+
+
+class ComparingRuleCondition(TypedDict):
+    """RuleCondition that compares a named field to a reference value."""
+
+    op: Literal["eq", "gt", "gte", "lt", "lte", "glob"]
+    name: str
+    value: Any
+
+
+class LogicalRuleCondition(TypedDict):
+    """RuleCondition that applies a logical operator to a sequence of conditions."""
+
+    op: Literal["and", "or"]
+    inner: Sequence[RuleCondition]
+
+
+class NotRuleCondition(TypedDict):
+    """RuleCondition that negates an inner condition."""
+
+    op: Literal["not"]
+    inner: RuleCondition
+
+
+class TagSpec(TypedDict):
+    """
+    Configuration for a tag to add to a metric.
+
+    Tags values can be static if defined through `value` or dynamically queried
+    from the payload if defined through `field`. These two options are mutually
+    exclusive, behavior is undefined if both are specified.
+    """
+
+    key: str
+    field: NotRequired[str]
+    value: NotRequired[str]
+    condition: NotRequired[RuleCondition]
+
+
+class MetricSpec(TypedDict):
+    """
+    Specification for a metric to extract from some data.
+
+    The metric type is given as part of the MRI (metric reference identifier)
+    which must follow the form: `<type>:<namespace>/<name>@<unit>`.
+
+    How the metric's value is obtained depends on the metric type:
+     - Counter metrics are a special case, since the default product counters do
+       not count any specific field but rather the occurrence of the event. As
+       such, there is no value expression, and the field is set to `None`.
+       Semantics of specifying remain undefined at this point.
+     - Distribution metrics require a numeric value.
+     - Set metrics require a string value, which is then emitted into the set as
+       unique value. Insertion of numbers and other types is undefined.
+    """
+
+    category: Literal["transaction"]
+    mri: str
+    field: NotRequired[Optional[str]]
+    condition: NotRequired[RuleCondition]
+    tags: NotRequired[Sequence[TagSpec]]
+
+
+def convert_alert_to_metric(snuba_query: SnubaQuery) -> Optional[MetricSpec]:
+    if snuba_query.dataset != Dataset.PerformanceMetrics.value:
+        return None
+
+    try:
+        spec = OndemandMetricSpec.parse(snuba_query.aggregate, snuba_query.query)
+        if not spec:
+            return None
+
+        return {
+            "category": DataCategory.TRANSACTION.api_name(),
+            "mri": spec.mri,
+            "field": spec.field,
+            "condition": spec.condition(),
+            "tags": [{"key": QUERY_HASH_KEY, "value": spec.query_hash()}],
+        }
+    except Exception as e:
+        logger.error(e, exc_info=True)
+        return None
+
+
+@dataclass(frozen=True)
+class OndemandMetricSpec:
```

Comment:
Let's add doc strings to all public functions and types in this module, describing their use and behavior in depth. While we will change a lot of the internals and tweak the API, we can already write down the intended design.